src/JWTVerifier.php

```diff
@@ -177,53 +178,35 @@ public function verifyAndDecode($jwt)
         }
 
         try {
-            $head_decoded = $this->decodeTokenSegment($tks[0]);
             $body_decoded = $this->decodeTokenSegment($tks[1]);
         } catch (\DomainException $e) {
             throw new InvalidTokenException('Malformed token.');
         }
 
-        if (! is_object($head_decoded) || ! is_object($body_decoded)) {
-            throw new InvalidTokenException('Malformed token.');
-        }
-
-        if (empty($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not found');
-        }
-
-        if (! $this->supportsAlg($head_decoded->alg)) {
-            throw new InvalidTokenException('Token algorithm not supported');
-        }
-
         // Validate the token audience, if present.
         if (! empty($body_decoded->aud)) {
             $audience = is_array($body_decoded->aud) ? $body_decoded->aud : [$body_decoded->aud];
             if (! count(array_intersect($audience, $this->valid_audiences))) {
-                $message  = 'Invalid token audience '.implode( ', ', $audience );
-                $message .= '; expected '.implode( ', ', $this->valid_audiences );
-                throw new InvalidTokenException($message);
+                throw new InvalidTokenException('Invalid token audience');
             }
         }
 
-        if ('HS256' === $head_decoded->alg) {
-            $secret = $this->client_secret;
+        // At this point, the constructor has set the supported algs to ensure only one.
+        if ($this->supportsAlg( 'HS256' )) {
+            $signature_key = $this->client_secret;
         } else {
```

Comment:
**for the scope of this function**, it could have received an array with more values than just HS256 and RS256. Instead of doing "else" I'd rather you check for explicit "RS256" which is the remaining case, and eventually fail in the else case at the end. Alternatively, at the top of this function write a check that the array doesn't contain anything other than RS256 or HS256.

Reply:
I agree. Performing explicit checks and failing unexpected values is better.

Reply:
The logic in this class is currently a bit hard to follow (mainly to try to avoid more breaking changes) but it already checks that the array only contains HS256 and RS256. So the else here could only mean that it's RS256. The constructor ensures that we only have a single value in there, HS256 if that value is in the supported algs passed (which throws an exception if there is no client secret) or RS256.